R/install_reqs.R

```diff
@@ -0,0 +1,174 @@
+    }
+  }
+  return(failures)
+}
+
+install_local = function(pkg){
+  #' @param pkg the local file to be installed
+  #' @return NULL
+  type = 'source'
+  if(!grepl('\\.tar\\.gz$',pkg)){
+    type = 'binary'
+  }
+  install.packages(pkg,repos=NULL,type=type)
+  return(NULL)
+}
+
+install_unversioned = function(elem, installed, dryrun, verbose, repo){
+  #' @param elem the list element of unversioned package requirements
+  #' @param installed list of installed packages
+  #' @param dryrun if TRUE, no packages will be installed, but you can see what would have happened
+  #' @param verbose should the function be explicit about activities
+  #' @param repo what CRAN repository should be used?
+  #'
+  #' @return the number of failures
+  failures = 0
+
+  if(length(elem) > 0){
+    for(i in elem){
+      if(is.na(installed[i])){
+        available_versions = get_available_versions(i)
+        v = vmess(sprintf(NOT_INSTALLED,i),verbose)
+        version = tail(sort(available_versions), 1)
+        v = vmess(sprintf(INSTALL,i,sprintf(INSTALL_VERSION,version)),verbose)
+        if(!dryrun){
+          tryCatch(install.packages(i,repos=repo),
+                   error=function(x){
+                     failures <<- failures + 1
+                     message(sprintf(OTHER_FAIL,i))
+                   })
+        }
+      }
+    }
+  }
+
+  return(failures)
+}
+
+is_versioned_install_needed = function(req, installed, verbose){
+  install_needed = FALSE
+  comp = sub(paste0('^.*?(',paste(COMPS,collapse='|'),').*$'),'\\1',req)
+  if(comp == "=") comp = "=="
+  split = strsplit(req,paste(COMPS,collapse='|'))[[1]]
+  package = gsub(' *','',split[1])
+  version = gsub(' *','',split[2])
+  if(is.na(installed[req])){
+    v = vmess(sprintf(NOT_INSTALLED,req),verbose)
+    install_needed = TRUE
+  } else {
+    install_needed = !check_version(version,installed[i],comp)
+  }
+  return(install_needed)
+}
+
+install_versioned = function(elem, installed, dryrun, verbose, repo){
+  #' @param elem the list element of unversioned package requirements
+  #' @param installed list of installed packages
+  #' @param dryrun if TRUE, no packages will be installed, but you can see what would have happened
+  #' @param verbose should the function be explicit about activities
+  #' @param repo what CRAN repository should be used?
+  #'
+  #' @return the number of failures
+  failures = 0
+
+  if(length(elem) > 0){
+    for(i in elem){
+      install_needed = is_versioned_install_needed(i, installed, verbose)
+      if(install_needed){
+        if(!is.na(installed[i])) v = vmess(sprintf(BAD_VERSION,package,installed[i]),verbose)
+
+        available_versions = get_available_versions(package)
+        available_compatibility = vapply(available_versions,
+                                         check_version,
+                                         TRUE,
+                                         target = version,
+                                         comp = comp)
```

Comment:
Where is `comp` coming from here?  I couldn't easily find it when looking at other `no visible binding for global variable` NOTES from the `devtools::check`

Reply:
I'll figure that out. Thanks for a thorough review

Reply:
This has been addressed in my most recent commit